source/find_getmore_killcursors_commands.rst

```diff
@@ -365,18 +365,35 @@ More detailed information about the interaction of the **secondaryOk** with **OP
 Behavior of Limit, skip and batchSize
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
-The new **find** command has different semantics to the existing 3.0 and earlier **OP_QUERY** wire protocol message. The **limit** field is a hard limit on the total number of documents returned by the cursor no matter what **batchSize** is provided.
-
-Once the limit on the cursor has been reached the server will destroy the cursor and return a **cursorId** of **0** in the **OP_REPLY**. This differs from existing **OP_QUERY** behavior where there is no server side concept of limit and where the driver **MUST** keep track of the limit on the client side and **MUST** send a **OP_KILL_CURSORS** wire protocol message when it limit is reached.
-
-When setting the **batchSize** on the **find** and **getMore** command the value MUST be based on the cursor limit calculations specified in the `CRUD`_ specification. 
-
-In the following example the **limit** is set to **4** and the **batchSize** is set to **3** the following commands are executed.
+The new **find** command has different semantics to the existing 3.0 and earlier
+**OP_QUERY** wire protocol message. The **limit** field is a hard limit on the
+total number of documents returned by the cursor no matter what **batchSize** is
+provided. This includes other limiting operations, such as the **$limit**
+aggregation pipeline stage. This differs from existing **OP_QUERY** behavior
+where there is no server side concept of limit and where the driver **MUST**
+keep track of the limit on the client side and **MUST** send a
+**OP_KILL_CURSORS** wire protocol message when it limit is reached.
+
+When setting the **batchSize** on the **find** and **getMore** command the value
+MUST be based on the cursor limit calculations specified in the `CRUD`_
+specification.
+
+Note that with 5.0, the server-side handling of cursors with a limit has
+changed. Before 5.0, some cursors were automatically closed when the limit was
+reached (e.g. when running **find** with **limit**), and the reply document did
+not include a cursor ID. Starting with 5.0, all cursor-producing operations will
+return a cursor ID if the end of the batch being returned lines up with the
+limit on the cursor. In this case, drivers MUST send an additional **getMore**
+request with **batchSize** 1 to have the cursor closed on the server side.
```

Comment:
Is the `batchSize` relevant here? Presumably _any_ further iteration would result in the server closing the cursor. We just need this to overcome the new query engine's behavior that avoids peaking ahead.

Reply:
Strictly speaking the batchSize is completely irrelevant here as long as we're fetching at least one result. I used `batchSize` 1 because that's what libmongoc uses, but I'm happy to accept any `batchSize` for the extra batch.

Reply:
Ok, I don't feel strongly. It might be awkward to have to split this up (e.g. MUST got `getMore`, SHOULD for a particular batchSize). Perhaps: "MUST send an additional `getMore` with a positive number for `batchSize` (e.g. 1)."

Reply:
I believe that pymongo will send a killCursors command in this case to cleanup the cursor which works on all server versions. Adding a requirement that we use getMore with batchSize=1 instead adds extra complexity since we'll still need to use killCursors on pre-3.2 servers. If we use the getMore batchSize=1 approach on older servers we could actually get 1 document which is wasteful, especially if the document is large.

Can we simplify this by always sending killCursors when the limit is reached instead of sometimes killCursors and sometimes getMore?

Reply:
As I said on another comment thread, I would prefer that this PR not require _any_ changes to any driver's cursor implementation.

Reply:
Yes I would prefer no driver changes as well.

Reply:
I've updated this to no longer prescribe a single method. The spec now requires drivers (the language uses "MUST", but I have no objections to change this to "SHOULD") to either use `getMore` or `killCursors` to close the server-side cursor if a cursor ID is returned.

I've also used this to clarify what the driver should do if the additional `getMore` (unexpectedly) yields more documents in the batch or still includes a cursor ID. I expect drivers to have varying implementations here and we don't test for this yet, so I'm fine with drivers not complying with this yet, especially since that is a very rare scenario (the driver would have to assume there's a limit on the cursor which the server doesn't see). I thought it was a good idea to be defensive here and provide this guidance.